python/ray/serve/async_goal_manager.py

```diff
@@ -8,34 +8,48 @@
 
 
 class AsyncGoalManager:
+    """
+    Class to help ServeController to manage desirable states of replica.
+
+    Essentially a dictionary of <goal_id, asyncio.Future> within
+    ServeController's event loop with APIs to facilitate common calls.
+    """
+
     def __init__(self):
-        self._pending_goals: Dict[GoalId, asyncio.Event] = dict()
+        self._pending_goals: Dict[GoalId, asyncio.Future] = dict()
 
     def num_pending_goals(self) -> int:
         return len(self._pending_goals)
 
     def create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
         if goal_id is None:
             goal_id = uuid4()
-        self._pending_goals[goal_id] = asyncio.Event()
+
+        self._pending_goals[
+            goal_id] = asyncio.get_running_loop().create_future()
         return goal_id
 
     def complete_goal(self, goal_id: GoalId) -> None:
         logger.debug(f"Completing goal {goal_id}")
-        event = self._pending_goals.pop(goal_id, None)
-        if event:
-            event.set()
+        future = self._pending_goals.pop(goal_id, None)
+        if future:
+            future.set_result(goal_id)
 
     def check_complete(self, goal_id: GoalId) -> bool:
+        # TODO: Maybe we can switch this to fut.done() instead ?
         return goal_id not in self._pending_goals
 
     async def wait_for_goal(self, goal_id: GoalId) -> None:
         start = time.time()
         if goal_id not in self._pending_goals:
             logger.debug(f"Goal {goal_id} not found")
             return
```

Comment:
In this case, even if the goal was errored, we would just return nothing. That seems OK for the corner case, but it's worth leaving a comment about this behavior.